File: apps/mo2-android/test/deep-link.test.ts
import { describe, it, expect } from "vitest";
import { parseDeepLink } from "../src/deep-link";

describe("parseDeepLink — App Links canonical + dub:// fallback (§2-4)", () => {
  it("App Link /home -> home", () => {
    expect(parseDeepLink("https://developershub.jp/home")).toEqual({ screen: "home" });
  });
  it("App Link event detail", () => {
    expect(parseDeepLink("https://developershub.jp/events/evt_123")).toEqual({
      screen: "eventDetail",
      eventId: "evt_123",
    });
  });
  it("App Link task detail", () => {
    expect(parseDeepLink("https://developershub.jp/tasks/tsk_9")).toEqual({
      screen: "taskDetail",
      taskId: "tsk_9",
    });
  });
  it("App Link inbox", () => {
    expect(parseDeepLink("https://developershub.jp/inbox")).toEqual({ screen: "inbox" });
  });

  it("dub:// fallback event detail", () => {
    expect(parseDeepLink("dub://events/evt_123")).toEqual({
      screen: "eventDetail",
      eventId: "evt_123",
    });
  });
  it("dub:// fallback home", () => {
    expect(parseDeepLink("dub://home")).toEqual({ screen: "home" });
  });
  it("dub:// fallback task", () => {
    expect(parseDeepLink("dub://tasks/tsk_1")).toEqual({ screen: "taskDetail", taskId: "tsk_1" });
  });

  it("retired devhub:// scheme -> unknown", () => {
    expect(parseDeepLink("devhub://home")).toEqual({ screen: "unknown", raw: "devhub://home" });
  });
  it("wrong host -> unknown", () => {
    const raw = "https://evil.example.com/tasks/tsk_1";
    expect(parseDeepLink(raw)).toEqual({ screen: "unknown", raw });
  });
  it("event without id -> unknown", () => {
    expect(parseDeepLink("https://developershub.jp/events")).toEqual({
      screen: "unknown",
      raw: "https://developershub.jp/events",
    });
  });
  it("garbage -> unknown", () => {
    expect(parseDeepLink("not a url")).toEqual({ screen: "unknown", raw: "not a url" });
  });

  // ---- S11 gantt ----
  it("App Link /events/{id}/gantt -> eventGantt", () => {
    expect(parseDeepLink("https://developershub.jp/events/evt_1/gantt")).toEqual({
      screen: "eventGantt",
      eventId: "evt_1",
    });
  });
  it("App Link /gantt/{eventId} shorthand -> eventGantt", () => {
    expect(parseDeepLink("https://developershub.jp/gantt/evt_2")).toEqual({
      screen: "eventGantt",
      eventId: "evt_2",
    });
  });
  it("dub://gantt/{eventId} fallback -> eventGantt", () => {
    expect(parseDeepLink("dub://gantt/evt_3")).toEqual({ screen: "eventGantt", eventId: "evt_3" });
  });
  it("gantt without eventId -> unknown", () => {
    expect(parseDeepLink("dub://gantt")).toEqual({ screen: "unknown", raw: "dub://gantt" });
  });
  it("plain /events/{id} still resolves to eventDetail (not gantt)", () => {
    expect(parseDeepLink("https://developershub.jp/events/evt_9")).toEqual({
      screen: "eventDetail",
      eventId: "evt_9",
    });
  });

  // ---- S10 chat ----
  it("App Link /chat -> chat list", () => {
    expect(parseDeepLink("https://developershub.jp/chat")).toEqual({ screen: "chat" });
  });
  it("App Link /chat/channels/{id} -> chatChannel", () => {
    expect(parseDeepLink("https://developershub.jp/chat/channels/chn_1")).toEqual({
      screen: "chatChannel",
      channelId: "chn_1",
    });
  });
  it("dub://chat fallback -> chat list", () => {
    expect(parseDeepLink("dub://chat")).toEqual({ screen: "chat" });
  });
  it("dub://chat/{id} shorthand -> chatChannel", () => {
    expect(parseDeepLink("dub://chat/chn_2")).toEqual({ screen: "chatChannel", channelId: "chn_2" });
  });
  it("chat channels without id -> unknown", () => {
    const raw = "https://developershub.jp/chat/channels";
    expect(parseDeepLink(raw)).toEqual({ screen: "unknown", raw });
  });
});

File: apps/mo2-android/android/app/src/test/kotlin/jp/developershub/dub/mo2/DeepLinkTest.kt
// Deep-link parser unit tests (§2-4) — App Links canonical + dub:// fallback,
// wrong host / scheme / missing id / malformed all resolve to Unknown.
// Mirrors deep-link.test.ts.
package jp.developershub.dub.mo2

import jp.developershub.dub.mo2.navigation.Route
import jp.developershub.dub.mo2.navigation.parseDeepLink
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class DeepLinkTest {

    @Test
    fun `app link home`() {
        assertEquals(Route.Home, parseDeepLink("https://developershub.jp/home"))
    }

    @Test
    fun `app link task detail`() {
        assertEquals(Route.TaskDetail("tsk_1"), parseDeepLink("https://developershub.jp/tasks/tsk_1"))
    }

    @Test
    fun `app link event detail`() {
        assertEquals(Route.EventDetail("evt_9"), parseDeepLink("https://developershub.jp/events/evt_9"))
    }

    @Test
    fun `app link inbox`() {
        assertEquals(Route.Inbox, parseDeepLink("https://developershub.jp/inbox"))
    }

    @Test
    fun `custom scheme fallback resolves the same routes`() {
        assertEquals(Route.Home, parseDeepLink("dub://home"))
        assertEquals(Route.TaskDetail("tsk_2"), parseDeepLink("dub://tasks/tsk_2"))
        assertEquals(Route.EventDetail("evt_2"), parseDeepLink("dub://events/evt_2"))
        assertEquals(Route.Inbox, parseDeepLink("dub://inbox"))
    }

    @Test
    fun `wrong host is unknown`() {
        assertTrue(parseDeepLink("https://evil.example.com/tasks/1") is Route.Unknown)
    }

    @Test
    fun `retired scheme is unknown`() {
        assertTrue(parseDeepLink("devhub://home") is Route.Unknown)
    }

    @Test
    fun `missing id is unknown`() {
        assertTrue(parseDeepLink("https://developershub.jp/tasks") is Route.Unknown)
        assertTrue(parseDeepLink("dub://events") is Route.Unknown)
    }

    @Test
    fun `unknown head is unknown`() {
        assertTrue(parseDeepLink("https://developershub.jp/settings") is Route.Unknown)
    }

    @Test
    fun `malformed input is unknown`() {
        assertTrue(parseDeepLink("not a url") is Route.Unknown)
        assertTrue(parseDeepLink("") is Route.Unknown)
    }
}
